model/integration_action_test.go

```diff
@@ -155,3 +156,120 @@ func TestPostActionIntegrationEquals(t *testing.T) {
 		require.False(t, pa1.Equals(pa2))
 	})
 }
+
+type validable interface {
+	IsValid() error
+}
+
+type validableTestCase struct {
+	name      string
+	validable validable
+	hasError  bool
+}
+
+func TestOpenDialogRequestIsValid(t *testing.T) {
+	validDialog := Dialog{CallbackId: "callback id", Title: "Test", IntroductionText: "introduction"}
+	invalidDialog := Dialog{}
+
+	testCases := []validableTestCase{
+		{name: "trigger id empty", validable: &OpenDialogRequest{TriggerId: "", URL: "http://mattermost.com", Dialog: validDialog}, hasError: true},
+		{name: "url empty", validable: &OpenDialogRequest{TriggerId: "trigger id", URL: "", Dialog: validDialog}, hasError: true},
+		{name: "url malformed", validable: &OpenDialogRequest{TriggerId: "trigger id", URL: "mattermost.com", Dialog: validDialog}, hasError: true},
+		{name: "invalid dialog", validable: &OpenDialogRequest{TriggerId: "trigger id", URL: "http://mattermost.com", Dialog: invalidDialog}, hasError: true},
+		{name: "all valid", validable: &OpenDialogRequest{TriggerId: "trigger id", URL: "http://mattermost.com", Dialog: validDialog}, hasError: false},
+	}
+
+	validableTestcasesRun(t, testCases)
+}
+
+func TestDialogIsValid(t *testing.T) {
+	makeValidDialogElementArray := func(nb int) (elements []DialogElement) {
+		for i := 0; i < nb; i++ {
+			elements = append(elements, DialogElement{DisplayName: "display", Name: fmt.Sprintf("Name %d", i), Type: "text", SubType: "password", Default: "default", Placeholder: "placeholder", HelpText: "help text", Optional: false, MinLength: 1, MaxLength: 150, DataSource: "", Options: nil})
+		}
+		return elements
+	}
+	invalidDialogElement := DialogElement{}
+
+	testCases := []validableTestCase{
+		{name: "callback id empty", validable: Dialog{CallbackId: "", Title: "title", IntroductionText: "intro text", IconURL: "http://mattermost.com/icon.svg", Elements: makeValidDialogElementArray(1)}, hasError: true},
```

Comment:
If we verify the error has the correct error message, we can assert that this specific validation is the one that failed. @JulienTant What do you think about this? We can have the test name match the error message, and keep the `hasError` boolean.

Also, we can create a helper function to clear up this test data. This makes it much easier to review the tests, including in the future when this code is revisited:

```go
func makeTestDialog(dialogIn Dialog, dependencies... []string) (Dialog, error) {
    dialog := Dialog{
        CallbackId: "callbackid",
        Title: "title",
        IntroductionText: "intro text",
        IconURL: "http://mattermost.com/icon.svg",
        Elements: makeValidDialogElementArray(1),
    }
    
    for _, field := range dependencies {
        switch field {
            case "Title":
                d.Title = dialogIn.Title
            case "CallbackId"
                // ...
            default:
                return dialog, errors.Errorf("unrecognized field name: %s", field)
        }
    }
    
    return dialog, nil
}
```

Then we can use it like so:

```go
testCase := {name: "callback id empty", validable: makeTestDialog(Dialog{CallbackId: ""}, "CallbackId")}
```

Reply:
Hi @mickmister thanks for the feedback! That's a good idea, let me work on those improvements.